File: leetcode/42-trapping-rain-water/main.go
package main

import "fmt"

/*
	1st approach:
	1. go forward to record the max among previous items
	2. go backward to record the max among previous items
	3. calculate the trap volumn for each index

	forward = 	[0 0 1 1 2 2 2 2 3 3 3 3]
	backward =  [3 3 3 3 3 3 3 2 2 2 1 0]
	result =		[0 0 1 0 1 2 1 0 0 1 0 0]
	result = 		1+1+2+1+1 = 6

	32 ms, faster than 15.73%
*/
func trap(height []int) int {
	// go forward to record the max among previous items
	max := 0
	forward := []int{}
	for i := 0; i < len(height); i++ {
		h := height[i]
		forward = append(forward, max)
		max = findMax(max, h)
	}
	// go backward to record the max among previous items
	max = 0
	backward := []int{}
	for i := len(height) - 1; i >= 0; i-- {
		h := height[i]
		backward = append([]int{max}, backward...)
		max = findMax(max, h)
	}
	// calculate the trap volumn for each index
	res := 0
	for i := 0; i < len(height); i++ {
		f := forward[i]
		b := backward[i]
		m := findMin(f, b)
		if m > height[i] {
			res += m - height[i]
		}
	}
	return res
}

func findMax(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func findMin(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(trap([]int{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}))
}

File: leetcode/42-trapping-rain-water/main.py
"""
    1st approach: 2pointers
    - similar to lc42, 135, 487, 689, 915, 1493
    - calculate max from the front
    - calculate max from the end
    - the min(forward[i], backward[i]) - height[i] is the volumn of the trapping water on that cell

    Time    O(3n)
    Space   O(3n)
    48 ms, faster than 35.09%
"""


class Solution(object):
    def trap(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        leftMax = 0
        forward = []
        for i in range(len(nums)):
            leftMax = max(leftMax, nums[i])
            forward.append(leftMax)
        backward = []
        rightMax = 0
        for i in range(len(nums)-1, -1, -1):
            rightMax = max(rightMax, nums[i])
            backward.append(rightMax)
        backward = backward[::-1]
        res = 0
        for i in range(len(nums)):
            res += min(forward[i], backward[i]) - nums[i]
        return res


print(Solution().trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]))

print("--------------------")


"""
    2nd approach: optimize the 1st approach
    - calculate max from the front
    - calculate max from the end
    - the min(forward[i], backward[i]) - height[i] is the volumn of the trapping water on that cell

    ref:
    - https://www.youtube.com/watch?v=2LjNzbK2cmA

    Time    O(n)
    Space   O(1)
    40 ms, faster than 43.69%
"""


class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        res = 0
        left = 0
        right = len(height) - 1
        leftMax = 0
        rightMax = 0
        while left < right:
            if height[left] < height[right]:
                leftMax = max(leftMax, height[left])
                res += leftMax - height[left]
                left += 1
            else:
                rightMax = max(rightMax, height[right])
                res += rightMax - height[right]
                right -= 1
        return res


print(Solution().trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]))
